File: AOC_2023/Day 8/8.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <map>
#include <numeric>

using namespace std;

class Solution
{
    public:
    string directions;
    map<string, pair<string,string>> nodes;
    void init(ifstream& input){
        string line;
        while (getline(input, line))
        {
            if(line.find("=") != string::npos){
                string node = line.substr(0, line.find(" ="));
                string left = line.substr(line.find("(")+1, 3);
                string right = line.substr(line.find(")")-3, 3);
                nodes[node] = make_pair(left, right);
            }
            else{
                if(directions.empty())
                    directions = line;
            }
        }

    }
    void part1(){
        int result = 0, n = directions.size();
        string start = "AAA", end = "ZZZ";
        while(start != end){
            if(directions[result%n] == 'L'){
                start = nodes[start].first;
            }
            else{
                start = nodes[start].second;
            }
            result++;
        }
        cout<<"Part 1: "<<result<<"\n";
    }
    void part2(){
        long long result = 1;
        int n = directions.size();
        vector<string> path;
        for(auto it : nodes){
            if(it.first[2] == 'A')
                path.push_back(it.first);
        }
        vector<long long> shortestPath(path.size());
        for(int i = 0; i < path.size(); i++){
            int cnt  = 0;
            while(true){
                if(directions[cnt%n] == 'L'){
                    path[i] = nodes[path[i]].first;
                }
                else{
                    path[i] = nodes[path[i]].second;
                }
                cnt++;
                if(path[i][2] == 'Z')
                    break;
            }
            shortestPath[i] = cnt;
        }
        for(auto it : shortestPath){
            result = lcm(result, it);
        }
        cout<<"Part 2: "<<result<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2023/Day 8/8.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

func initialize(filename string) (string, map[string][]string, error) {
	file, err := os.Open(filename)
	if err != nil {
		return "", nil, err
	}
	defer file.Close()
	var directions string
	nodes := make(map[string][]string)
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		text := scanner.Text()
		if strings.Contains(text, "=") {
			var temp []string
			node := text[0: strings.Index(text, " =")]
			temp = append(temp, text[strings.Index(text, "(") + 1: strings.Index(text, ",")] )
			temp = append(temp, text[strings.Index(text, ",") + 2: strings.Index(text, ")")] )
			nodes[node] = temp
			
		} else {
			if directions == "" {
				directions = scanner.Text()
			}
		}	
	}
	if err := scanner.Err(); err != nil {
		return "", nil, err
	}
	return directions, nodes, nil
}

func gcd(a, b int64) int64 {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func lcm(a, b int64) int64 {
	return a * b / gcd(a, b)
}

func part1(directions string, nodes map[string][]string) {
	result := 0
	n := len(directions)
	start := "AAA"
	end := "ZZZ"
	for start != end {
		if directions[result%n] == 'L' {
			start = nodes[start][0]
		} else{
			start = nodes[start][1]
		}
		result++
	}
	fmt.Println("Part 1:", result)
}

func part2(directions string, nodes map[string][]string) {
	result := int64(1)
	n := len(directions)
	var startpos []string
	for key := range nodes {
		if key[2] == 'A'{
			startpos = append(startpos, key)
		}
	}
	for _, start := range startpos {
		cnt := int64(0)
		for start[2] != 'Z' {
			if directions[int(cnt)%n] == 'L' {
				start = nodes[start][0]
			} else{
				start = nodes[start][1]
			}
			cnt++
		}
		result = lcm(result, cnt)
	}
	fmt.Println("Part 2:", result)
}

func main(){
	filename := "input.txt"
	directions, nodes, err := initialize(filename)
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	part1(directions, nodes)
	part2(directions, nodes)
}